### DISCUSSION TITLE
Inverted relay outputs "blip" on cold boot

### DISCUSSION THREAD

**Discussion by numero-trey** [arendst/Tasmota#15630] [Q&A] [answered]
### PROBLEM DESCRIPTION

When cold booting an ESP8266 or ESP32, inverted relay outputs flash on and off quickly.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): ESP8266 & ESP32
- [x] Tasmota binary firmware version number used: 11.1.0.3
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: Web installer
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
Backlog Template; Module; GPIO 255
06:53:25.630 RSL: RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,256,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
06:53:25.837 RSL: RESULT = {"Module":{"0":"Generic"}}
06:53:26.092 RSL: RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"256":"Relay_i1"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}

# Additional config to disable saving output state, set outputs off at boot, and disable relay state feedback scan.
Backlog SetOption0 0; PowerOnState 0; SetOption63 0
```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
n/a
```
- [x] Provide the output of this command: `Status 0`:
```lua
06:54:29.026 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_70508F","ButtonTopic":"0","Power":0,"PowerOnState":0,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":0,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
06:54:29.052 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"External System","Uptime":"0T00:01:39","StartupUTC":"2022-05-15T05:52:50","Sleep":50,"CfgHolder":4617,"BootCount":9,"BCResetTime":"2022-05-15T05:01:55","SaveCount":22,"SaveAddress":"F4000"}}
06:54:29.084 RSL: STATUS2 = {"StatusFWR":{"Version":"11.1.0.3(tasmota)","BuildDateTime":"2022-05-14T15:42:39","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"368/699"}}
06:54:29.104 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["BikiniBottom_IoT",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008008","2805C80001000600003C5A0A190000000000","00000080","00006000","00004000"]}}
06:54:29.135 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":616,"Free":384,"Heap":27,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","04000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
06:54:29.165 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-70508F-4239","IPAddress":"192.168.15.194","Gateway":"192.168.15.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.15.1","DNSServer2":"0.0.0.0","Mac":"CC:50:E3:70:50:8F","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
06:54:29.189 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_70508F","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
06:54:29.212 RSL: STATUS7 = {"StatusTIM":{"UTC":"2022-05-15T05:54:29","Local":"2022-05-15T06:54:29","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"05:08","Sunset":"20:24"}}
06:54:29.230 RSL: STATUS10 = {"StatusSNS":{"Time":"2022-05-15T06:54:29"}}
06:54:29.237 RSL: STATUS11 = {"StatusSTS":{"Time":"2022-05-15T06:54:29","Uptime":"0T00:01:39","UptimeSec":99,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"BikiniBottom_IoT","BSSId":"E0:63:DA:A4:CF:E1","Channel":6,"Mode":"11n","RSSI":100,"Signal":-42,"LinkCount":1,"Downtime":"0T00:00:04"}}}
```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
 00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at FB, Count 23
00:00:00.055 QPC: Count 1
00:00:00.058 CFG: CR 368/699, Busy 0
00:00:00.061 ROT: Mode 1
00:00:00.062 SRC: Restart
00:00:00.064 Project tasmota - Tasmota Version 11.1.0.3(tasmota)-2_7_4_9(2022-05-14T15:42:39)
00:00:00.217 WIF: Checking connection...
00:00:00.218 WIF: Attempting connection...
00:00:00.554 WIF: Connecting to AP1 BikiniBottom_IoT Channel 6 BSSId E0:63:DA:A4:CF:E1 in mode 11n as tasmota-70508F-4239...
00:00:01.786 WIF: Checking connection...
00:00:01.787 WIF: Attempting connection...
00:00:02.802 WIF: Checking connection...
00:00:02.803 WIF: Connected
00:00:03.054 HTP: Web server active on tasmota-70508F-4239 with IP address 192.168.15.194
00:00:03.402 WIF: Sending Gratuitous ARP
00:00:03.403 NTP: Sync time...
00:00:03.535 RTC: UTC 2022-05-15T05:59:32, DST 2022-03-27T02:00:00, STD 2022-10-30T03:00:00
06:59:32.001 RTC: Synced by NTP
06:59:33.043 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"11.1.0.3(tasmota)","FallbackTopic":"cmnd/DVES_70508F_fb/","GroupTopic":"cmnd/tasmotas/"}}
06:59:33.049 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-70508F-4239","IPAddress":"192.168.15.194"}}
06:59:33.059 RSL: INFO3 = {"Info3":{"RestartReason":"External System","BootCount":10}}
06:59:33.067 RSL: RESULT = {"POWER":"OFF"}
06:59:33.070 RSL: POWER = OFF
06:59:35.465 QPC: Reset
06:59:37.373 APP: Boot Count 10
06:59:37.378 RSL: STATE = {"Time":"2022-05-15T06:59:37","Uptime":"0T00:00:10","UptimeSec":10,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"BikiniBottom_IoT","BSSId":"E0:63:DA:A4:CF:E1","Channel":6,"Mode":"11n","RSSI":100,"Signal":-41,"LinkCount":1,"Downtime":"0T00:00:04"}}
06:59:37.589 CFG: Saved to flash at FA, Count 24, Bytes 4096
06:59:42.408 RSL: config = {"ip":"192.168.15.194","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-70508F-4239","mac":"CC50E370508F","md":"Generic","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"11.1.0.3","t":"tasmota_70508F","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"sho":[0,0,0,0],"sht":[[0,0,0],[0,0,0],[0,0,0],[0,0,0]],"ver":1}
06:59:42.411 RSL: sensors = {"sn":{"Time":"2022-05-15T06:59:42"},"ver":1}
```

### TO REPRODUCE
_Steps to reproduce the behavior:_

1. Connect LED & resistor (or low-on relay) between GPIO pin and ground.
2. Configure output pin as Relay_i
3. Unplug power, and re-plug

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

Input should remain off as device boots.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
N/A

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

Tried on outputs that are pulled high at boot and inputs that are not.

Arduino sketch below shows expected behavior, so the blip is coming from the Tasmota init process. Also tried on ESP32 w/`gpio_reset_pin` that is monkey patched into pinMode.

```C
# Output pin D1
#define PIN 5
void setup() {
  pinMode(PIN, OUTPUT);
  digitalWrite(PIN, HIGH);
}

void loop() {
  delay(2000);
  digitalWrite(PIN, !digitalRead(PIN));
}
```

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Unplug and restore previous state without a blink of output is not supported.
How should this work? If there is no power for a longer time relay is without power.
If you mean it does blink after the command `restart 1` this would be not like it should be.
I have several smart plugs. None does blink when doing a firmware upgrade.

**Reply by numero-trey** [reply]
This is not for relays where I want the relay output inverted. This is an inverted GPIO controlling relay modules that need to be pulled low to activate. Please see my response below.

**Comment by numero-trey** [comment]
This issue is with relays that must be pulled low to activate (or the LED in my repro steps). This applies to most of the generic relay modules you can get for custom hardware.

When the ESP cold boots, the GPIOs are not pulled low. If the pin state is set before the pin mode (or immediately after), there is no blip (or it is effectively eliminated) as my demo sketch illustrates. The only way I could simulate the issue in my demo sketch was to add a delay between pinMode and digitalWrite. Tasmota initializes PWM_i outputs in a manner to not blip, so it is possible.

The fix for this would be to set the correct state before setting the pin as an output.

A warm restart (restart 1) does not trigger the blip. This is only during the Tasmota init process after a cold boot.

**Comment by s-hadinger** [comment]
Good point. Setting the output mode for Relays is done here:
https://github.com/arendst/Tasmota/blob/91724dc8c123a036ebd9cd4e140c6a63ffa4645e/tasmota/support_tasmota.ino#L2106-L2117

But the sequence is here: 
https://github.com/arendst/Tasmota/blob/91724dc8c123a036ebd9cd4e140c6a63ffa4645e/tasmota/tasmota.ino#L509-L526

So there can be a delay between `GpioInit()` and `SetPowerOnState()`

@arendst should we set the value to low/high during this period? It may interfere with a warm restart.

**Reply by numero-trey** [reply]
Thanks for all the detail! I was struggling to find the pinMode calls for relays.

It looks like support_tasmota.ino#2112 is a workaround for this issue on a specific piece of hardware.

I wonder if SetPowerOnState() is idempotent. I'm going to do a build calling it before and after (even though I have no way of validating that change on 2% of supported hardware).

**Comment by arendst** [comment]
- On an ESP8266 did you play with ``SetOption63``?
- Did you really observer the issue on ESP32?

**Reply by numero-trey** [reply]
Yes and yes. I'm using `SetOption0 0; PowerOnState 0; SetOption63 0` in my config. I initially found the blip on the ESP32 I'm prototyping with and confirmed it on an 8266 before opening the bug.

**Comment by vic42** [comment, marked as answer]
I verified this issue with an ESP32-Devkit-V4 (current dev version).
It does blink when a GPIO is defined as `Relay` or `Relay_i`. It does **not** blink on an unconfigured GPIO.
- A `Relay` does a short "OFF" blink when ON
- A `Relay_i` does a short "ON" blink when OFF

"short" means 1/2 second to 1 second
`SO63` does not cause a change

**Reply by arendst** [reply]
Investigating...

**Reply by arendst** [reply]
Try latest dev release. It moved the ``pinMode`` functions just above the ``digitalWrite`` fixing the "blip" I noticed too.

**Reply by vic42** [reply]
I can't reproduce the issue with the new version so I assume it's resolved.

**Reply by arendst** [reply]
Thx for the feedback.

**Comment by mahipat99** [comment]
I had made one prototype board with a few switches and a 4CH SSR(low level) board. Whenever I pressed the reset switch or power cycle, it had the same blip issue.
FW - 11.1.0
Board - Nodemcu ESP32-WROOM-32

After seeing the issue, I thought to recompile it with the latest commit(51af861), it seems solved for 2 relays (GPIO 15,16), but still, 2 relay has a problem connected to GPIO 2 & 4.

Template
`{"NAME":"ESP32-15PIN","GPIO":[0,0,256,0,257,0,0,0,0,0,165,258,259,0,1376,166,0,167,0,0,0,162,163,164,0,0,0,0,160,161,0,0,0,0,0,0],"FLAG":0,"BASE":1}`

`Backlog SetOption0 0; PowerOnState 0; SetOption63 1`

With simple Arduino code or ESPHome, it works without any issue.

**Reply by arendst** [reply]
I cannot reproduce. My esp32-devkit has a led connected to GPIO02 so connecting the relay too will indeed show some kind of blip for obvious reason. All other GPIO's do not do any blip.

Instead of using release 11.1.0 I suggest you use development 11.1.0.3 which contains more fixes than just commit (https://github.com/arendst/Tasmota/commit/51af8611bbc461de5af5f4ff879b128602b7d4c6) which is only part of the solution. Download pre-compiled devs from http://ota.tasmota.com/tasmota32/

**Reply by vic42** [reply]
With ESP32-WROVER I observed a very short Low-blip for GPIOs that are High Relay_i (I haven't measured the actual duration but I guess it is less than a few milliseconds) happening early during cold boot on all GPIOs I checked. A relay will certainly not change state. However, if an SSR is very fast it might hand it through to the load. Practical implications should be very limited.

**Reply by mahipat99** [reply]
> I cannot reproduce. 

To double-check, I've again tried Arduino code, and ESPHome both have the same issue. Maybe because GPIO 2 and 4 are strapping pins or have an internal pull-down resistor.
I guess I had tried other pins earlier, but I can't recall it. Sorry for the confusion.

Rest pins issue solved, as I stated in an earlier comment. Thank you very much.

**Comment by numero-trey** [comment]
I can confirm my relays (and their status LEDs) are not blipping with 11.1.0.3.

Thanks for the fix, @arendst , and @s-hadinger , @vic42 , and @mahipat99 for helping!

**Reply by maurimerel** [reply]
hi there , Iḿ having the same issue as yours, i have an esp01 connected to a two relay module , outputs were set with pulsetime 20 , they control two gates . Everything works great until it cold boots , relays "blip" briefly when booting and my gates are opened . 
Two years have passed and thought your solution was included in later releases . Itś not clear to me how you solved it and cant find the 11.0.1.3 version to take a look . Can you guide me a bit ??? thanks in advance!!!

**Reply by s-hadinger** [reply]
@maurimerel It's hard to guess which version you are running and why you want specifically 11.1.0.3. Please try with the latest release v12.1.1 and report back.

**Reply by mahipat99** [reply]
@maurimerel Mention pins too, or check pins here that are high on the boot
https://randomnerdtutorials.com/esp8266-pinout-reference-gpios/

**Reply by sfromis** [reply]
On the ESP-01 (pretty unsuitable device for builds), you cannot avoid pin effects at boot, as all 4 available gpio pins are either output high at boot (gpio 1,2,3), as defined by ESP8266 specs, or (gpio 0) **must** be pulled high at boot.

**Reply by vic42** [reply]
See also https://rabbithole.wwwdotorg.org/2017/03/28/esp8266-gpio.html for a discussion on ESP8266 GPIO cold boot behavior. An ESP01 does not expose GPIO pins on its header that stay low during cold boot. This has nothing to do with Tasmota, let alone the original issue.

**Reply by sfromis** [reply]
Technically, they do not stay floating either, they go high, before Tasmota gets control.

**Reply by maurimerel** [reply]
Thanks to you all guys , you were right ESP01 is not suitable for this case , read the articles mentioned by @mahipat99 and @vic42 , tried with a NodeMcu board and now everything works just right . Learning everyday , never came to my mind that those pins had such different behaviours . 

Thanks again @s-hadinger @mahipat99 @sfromis @vic42 .
